### ISSUE TITLE
Accessibility on login page & Password Manager Support

### ISSUE THREAD

**Issue by creack** [home-assistant/frontend#3133]
Without going in the depth of WCAG & co, it would be great if we could have a little better accessibility, especially on the login page.

While it looks nice, the login/password fields are defined as "paper input 1" and "paper input 2", which is a big no-no for accessibility, and in my case, prevents my password manager (LastPass, but I suspect it would be the same for others) to understand it is a login form and where it should populate the login/password, resulting in having to manually copy/paste the password each time.

**Comment by Nowaker** [comment]
Current:
```
<input autocomplete="off" placeholder="" autocapitalize="none" autocorrect="off" aria-describedby="paper-input-add-on-1" aria-labelledby="paper-input-label-1">
```
Requested:
```
<input name="username" placeholder="" autocapitalize="none" autocorrect="off" aria-describedby="paper-input-add-on-1" aria-labelledby="paper-input-label-1">
```

And `name="password"` for password. This will fix LastPass.

**Comment by miggs125** [comment]
I'd be happy to work on this if it's not taken.

**Comment by springstan** [comment]
@miggs125 any updates on this so far? :) Looks like a really great issue that I would also like to contribute to.

**Comment by miggs125** [comment]
@springstan I haven't made much progress I've been really busy lately. Feel free to take it!

**Comment by springstan** [comment]
@miggs125 thanks for the update! I will see what I can do :)

**Comment by springstan** [comment]
@bramkragten thanks for assigning me :)

I am currently going through the project trying to find the place where I would have to make changes.

`ha-authorize.ts` renders among other things the render of `ha-auth-flow.ts`. 
That render consists of a `form` which has among other things two `ha-form` as children.
One for username and another for the password. I am assuming that for both forms `ha-form-string.ts` is used for rendering.

With all of that I still cannot find the `input`s for username and password, not even its parent `iron-input`. Does anyone where they are rendered? :)

**Comment by bramkragten** [comment]
`ha-form-string` is rendering a `paper-input` the actual `input` element comes from `paper-input`, you should not touch that.

**Comment by bramkragten** [comment]
Btw what is stated in this issue is not correct `paper-input-add-on-1` and `paper-input-add-on-2` are correct, they refer to the `id` of the label of the `paper-input` and are labelled correct.

**Comment by springstan** [comment]
@bramkragten thanks again for the insight. I have researched what LastPass and other password managers require to detect a form as a login form. 
An example from the [LastPass page](https://support.logmeininc.com/lastpass):
```html
<form action="https://mypage.com/blah" method="post">
    <input type="text" name="username" id="username" value=""/>
    <input type="password" name="password" id="password" value=""/>
    <input type="submit" value="LOGIN"/>
</form>
```
So I have tried to replicate that, but LastPass does not seem to pick it up.

What I have so far:
- added action and method to form
- changed mwc button to input type=submit
- set the name attribute on username and password input

What I could not get to work:
- setting the id on username and password input

**Comment by bramkragten** [comment]
Oh, I thought I mentioned that; I don't think we can solve this for lastpass and other password managers because they don't support shadow dom...

And we don't use forms, so I wonder where you put the `action` and `method`?

**Comment by springstan** [comment]
Oh my mistake then, I thought this could be solved.
Yeah it is not actively used but it is there:
https://github.com/home-assistant/home-assistant-polymer/blob/3c0ba1d7eb27d49a2eae8003b15c7baec7ddbd01/src/auth/ha-auth-flow.ts#L33-L35

**Comment by springstan** [comment]
Could you unassign me? Since we cannot seem to solve this because we are using shadow dom.

**Comment by wwwouter** [comment]
> Oh, I thought I mentioned that; I don't think we can solve this for lastpass and other password managers because they don't support shadow dom...

Could we create 2 "real" inputs that are connected to the shadow DOM, but are hidden and only there for LastPass ea?

**Comment by bramkragten** [comment]
I'm not sure LastPass will work on hidden input fields, but if anyone want to try it...

**Comment by Nowaker** [comment]
LastPass fills fields type=text and type=password with id or name attributes only. I guess the login page should be a regular HTML login page if SPA makes the basics overly complicated.

**Comment by NdS-Research-Facilities** [comment]
This is quickly becoming a PITA,
What is required to add a new (standard) login page ?

**Comment by outadoc** [comment]
Bumping this issue because it's always been very painful to me, especially when working a lot with OAuth, or just setting up hass on new devices. Not much of a frontend dev but if someone has more time to look into I think it would be a very welcome contribution.

**Comment by Taomyn** [comment]
Please can we get this fixed - everyone is told they should be using password managers for extra security, so having a website that doesn't conform to allow passwords to be auto-filled doesn't help.

**Comment by zsarnett** [comment]
https://github.com/home-assistant/frontend/issues/1622

**Comment by Nowaker** [comment]
Can you guys upvote the top-level comment here? It will increase the visibility of this issue.

**Comment by zsarnett** [comment]
Reading about this in slack and how others have done this, it seems that the most common way is adding the hidden inputs to the Body and on changes to either input (on the body or in element) it updates the other.

https://github.com/home-assistant/frontend/blob/auth-password-manager/src/components/ha-form/ha-form-string.ts#L42

Something like this. right now it's one way but this may be the easiest way. A lot of tweaks needed, adding those fields in a form. Cleaning up the code, updating both ways, actual testing.

**Comment by mnoorenberghe** [comment]
An easier way would be not using Shadow DOM for the two inputs and wrapping them in a `<form>`. It may require duplicating some CSS but it will address the accessibility and password manager issues.

**Comment by hadim** [comment]
Same issue here. I can't save my password so I end up using a weak password I can remember instead of Chrome password manager. It makes my HA installation vulnerable.

**Comment by JeffResc** [comment]
In response to hiding the input elements, it seems that something like that is totally possible. Amazon's login page actually does this and it works flawlessly with my password manager, BitWarden.

![Screenshot](https://i.imgur.com/Rj8YGdT.png)

**Comment by Misiu** [comment]
Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1629226
This should be supported by the browser.
If anyone finds bug reports for other browsers post them here.

**Comment by bramkragten** [comment]
@JeffResc That Amazon page is not using shadowdom right?

**Comment by Nowaker** [comment]
Can we **simply** stop using the latest tech that is password filler and usability unfriendly and go back to normal HTML? We don't need the 2020 tech on the login page. 1995 tech sans the gradients and shadows will do...

**Comment by hadim** [comment]
Agree with the above I just want a standard login page that works with the Chrome Password Manager.

**Comment by zsarnett** [comment]
Clearly, yes we want to make the password manager work. Its not as easy and just a standard login  page with Just HTML. If you don't have anything helpful to comment please just add a thumbs up 😃 We understand this is a feature that is wanted

**Comment by Nowaker** [comment]
> Its not as easy and just a standard login page with Just HTML

That's the point! Let's make it one! :smile:

**Comment by JeffResc** [comment]
@Nowaker I’m not sure, it’s just the default login on Amazon.com, although I know enough HTML and JS to get around, I’m not sure about the specifics of Shadow DOM or how to determine if it’s in use.

**Comment by adriancampos** [comment]
> This should be supported by the browser.
> If anyone finds bug reports for other browsers post them here.

Here's a chromium bug report from 2018: https://bugs.chromium.org/p/chromium/issues/detail?id=885177
Unfortunately, it was marked as wontfix and the referenced site worked around the issue by using normal input elements.

**Comment by chasebolt** [comment]
Wild that this is still an open issue. If something like this is still open, it makes me wonder how secure is HomeAssistant Cloud.

**Comment by spacegaier** [comment]
> 
> 
> Wild that this is still an open issue. If something like this is still open, it makes me wonder how secure is HomeAssistant Cloud.

What does this have to do with the security of the HA cloud integration? Regardless: Free open-source project => feel free to tackle this ticket.

**Comment by chasebolt** [comment]
The mentality of being open source "you fix it", while the company also uses it to provide a paid for service - is a broken model.

If the priority of a company can't sort out getting the dominant password managers to work on its login page. How do you believe that doesn't translate to the hosted infrastructure in any way?

**Comment by NullEnt1ty** [comment]
> The mentality of being open source "you fix it", while the company also uses it to provide a paid for service - is a broken model.
> 
> If the priority of a company can't sort out getting the dominant password managers to work on its login page. How do you believe that doesn't translate to the hosted infrastructure in any way?

As someone who uses password managers regularly, I can relate to your frustration. But it's not like password managers are not working at all with Home Assistant. You can still copy&paste your credentials or use auto-typing (if this is a feature that is supported by your password manager). It's just a small inconvenience in my opinion, so I can see why it's not on top of the priority list.

**Comment by pfrenssen** [comment]
You cannot enforce a priority on an open source project anyway. It will only be fixed if someone cares so much about it that they are willing to put in the required work. The problem is that this is implemented in a pretty bizarre and nonstandard way and it is not so easy to fix in a few hours. The reason this is not fixed is simply because nobody cares enough to put in a lot of hours / days to unravel and fix it.

**Comment by timdonovanuk** [comment]
+1 but looking at the thread doesn't seem like a reality :|

**Comment by asafbrukarz** [comment]
Hi all, I was looking into this issue a little today. It looks like it was introduced with https://github.com/home-assistant/frontend/issues/228. The issue is with paper-input and it seems they're not interested in fixing this.

There's an issue with browsers supporting autofill with shadow DOM, and using shady DOM seemed to work in my tests. See the Stack Overflow comment [here](https://stackoverflow.com/a/48687882)

Switching to shady for the affected login pages seemed to work for me.

**Comment by Azertooth** [comment]
this problem is going from mildly infuriating to a total embarass, over 2 years without a fix

**Comment by Nowaker** [comment]
The login page should be a regular HTML login page if SPA makes the basics overly complicated.

**Comment by johanson** [comment]
The founder seems to suggest it wont' be fixed - https://github.com/home-assistant/frontend/issues/1622#issuecomment-577014115
I agree with @Nowaker, common sense would be the login page to work with password managers, not aiming for to be 100% x framework. Currently the Supervisor is scanning for weak/pwned passwords that you're using for addons (even if behind LAN) and nags you about them, but you can't use a password manager to login to HA itself. Doesn't make sense :)

**Comment by NullEnt1ty** [comment]
> but you can't use a password manager to login to HA itself.

That's not true. Browser integrations from most password managers won't work, but you can still copy&paste your credentials or use auto-typing as I've stated earlier.

I'm using KeePassXC on several platforms to log into Home Assistant.

**Comment by rianadon** [comment]
Continuing what @asafbrukarz said, I think enabling shady DOM for just the login page is the most pragmatic way to fix this. I was able to reproduce and using the shady dom polyfill allows password managers to see the fields.

This method isn't without performance issues, but for what we gain, I think it's a great tradeoff:
- The polyfill when sent over the network compressed adds an additional 30 KB
- On my old (6 year old but decently powerful) laptop in Chrome, using the polyfill delays the username & password form from showing up by 300 ms.

I'd be happy to create a PR if this tradeoff is acceptable.

**Comment by tablatronix** [comment]
This is a damn nightmare, and should be fixed at all costs.

**Comment by rianadon** [comment]
I'm making a list of the accessibility issues I found for my own reference. Please correct me if I'm missing anything:
- Home Assistant Logo has no alt text
- The alternate authentication method list is not properly described when using VoiceOver
- With the trusted networks authentication method, pressing tab while focusing the User dropdown does not advance to the NEXT button but instead keeps the dropdown focused
- On that same dropdown, it is impossible to navigate via the keyboard

The good (and bad) news is that most of these components are also used in the Lovelace dashboard, so fixing these means more wins for accessibility!

**Comment by tablatronix** [comment]
@sunsear Using Shadow DOM was the issue I believe, not something simple like that.

**Comment by rianadon** [comment]
@sunsear https://github.com/home-assistant/frontend/pull/9235 should have fixed issues with password managers (the solution is essentially bypassing MWC/Polymer/Shadow DOM and adding plain `input` elements to the root of the HTML tree).

If your password manger isn't working with autocomplete, I'd love to hear which one / browser you are using so we can debug this 🐛  But until either browsers & password mangers support shadow dom or MWC stops using shadow dom, the solution for now will be beefing up that `input` polyfill.

**Comment by creack** [comment]
Note that this is also an issue with Bitwarden.

**Comment by balloob** [comment]
Closing this issue as it works now. Please open a new issue if you encounter a new issue.

**Comment by Nowaker** [comment]
@balloob Please don't ignore the problems the community is still reporting with this and reopen this ticket.
